File: go/test/clone_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/bluefin-shieldconex-mgmt-sdk/go"
	"github.com/voxgig-sdk/bluefin-shieldconex-mgmt-sdk/go/core"

	vs "github.com/voxgig-sdk/bluefin-shieldconex-mgmt-sdk/go/utility/struct"
)

func TestCloneEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Clone(nil)
		if ent == nil {
			t.Fatal("expected non-nil CloneEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := cloneBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "clone." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		cloneRef01Ent := client.Clone(nil)
		cloneRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "clone"}, setup.data), "clone_ref01"))
		cloneRef01Data["template_id"] = setup.idmap["template01"]

		cloneRef01DataResult, err := cloneRef01Ent.Create(cloneRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		cloneRef01Data = core.ToMapAny(cloneRef01DataResult)
		if cloneRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if cloneRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

	})
}

func cloneBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "clone", "CloneTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read clone test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse clone test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"clone01", "clone02", "clone03", "template01", "template02", "template03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID": idmap,
		"BLUEFINSHIELDCONEXMGMT_TEST_LIVE":      "FALSE",
		"BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN":   "FALSE",
		"BLUEFINSHIELDCONEXMGMT_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["BLUEFINSHIELDCONEXMGMT_APIKEY"],
			},
			extra,
		})
		client = sdk.NewBluefinShieldconexMgmtSDK(core.ToMapAny(mergedOpts))
	}

	live := env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/clone_entity_test.rb
# Clone entity test

require "minitest/autorun"
require "json"
require_relative "../BluefinShieldconexMgmt_sdk"
require_relative "runner"

class CloneEntityTest < Minitest::Test
  def test_create_instance
    testsdk = BluefinShieldconexMgmtSDK.test(nil, nil)
    ent = testsdk.Clone(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = clone_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "clone." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    clone_ref01_ent = client.Clone(nil)
    clone_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.clone"), "clone_ref01"))
    clone_ref01_data["template_id"] = setup[:idmap]["template01"]

    clone_ref01_data_result = clone_ref01_ent.create(clone_ref01_data, nil)
    clone_ref01_data = Helpers.to_map(clone_ref01_data_result)
    assert !clone_ref01_data.nil?
    assert !clone_ref01_data["id"].nil?

  end
end

def clone_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "clone", "CloneTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = BluefinShieldconexMgmtSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["clone01", "clone02", "clone03", "template01", "template02", "template03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID" => idmap,
    "BLUEFINSHIELDCONEXMGMT_TEST_LIVE" => "FALSE",
    "BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN" => "FALSE",
    "BLUEFINSHIELDCONEXMGMT_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["BLUEFINSHIELDCONEXMGMT_TEST_CLONE_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["BLUEFINSHIELDCONEXMGMT_APIKEY"],
      },
      extra || {},
    ])
    client = BluefinShieldconexMgmtSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["BLUEFINSHIELDCONEXMGMT_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["BLUEFINSHIELDCONEXMGMT_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
